File: c/p039.c
#include <stdio.h>

int num_solutions(int p)
{
    int solutions = 0;
    for (long a = 2; a < (p/3); a++) 
    { 
        if (p*(p-2*a) % (2*(p-a)) == 0)
        { 
            solutions++; 
        } 
    }
    return solutions; 
}

int main()
{
    int result = 0;
    int result_solutions = 0;

    for (int p = 2; p <= 1000; p += 2) {
        int p_solutions = num_solutions(p);
        if (p_solutions > result_solutions)
        {
            result_solutions = p_solutions;
            result = p;
        }
    }
    printf("%d\n", result);
}

File: python/p039.py
def num_solutions(p):
    count = 0
    for a in range(2, p//3):
        # Know that b must be (p**2 - 2*a*p) / (2*(p - a))
        # by solving a**2 = b**2 = (p - a - b)**2
        # Therefore, for b to be an integer,
        #  (p**2 - 2*a*p) % (2*(p - a)) == 0
        if (p**2 - 2*a*p) % (2*(p - a)) == 0:
            count += 1
    return count

def compute():
    max_p = (0, 0) # (p, # of solutions)
    # If one of a or b is odd, c is odd => p is even
    # If both a and b are odd, c is even => p is even
    # If a and b are both even, c is even => p is even
    for p in range(12, 1001, 2):
        sols = num_solutions(p)
        if sols > max_p[1]:
            max_p = (p, sols)
    return max_p[0]

if __name__ == "__main__":
    print(compute())
